### PR TITLE
split dynamic linker representations from compilers

### LINKED ISSUE #4784: Meson not detecting Apple linker with standard Clang
We [wanted](https://github.com/NixOS/nixpkgs/pull/54007) to link `glib` against Mac libraries but the build failed with:

```
FAILED: gio/libgio-2.0.0.dylib
cc  -o gio/libgio-2.0.0.dylib 'gio/6ae6c9e@@gio-2.0@sha/meson-generated_.._gioenumtypes.c.o' […] 'gio/kqueue/9c7395d@@kqueue@sta/dep-list.c.o' -Wl,-O1 -shared -fPIC -Wl,-soname,libgio-2.0.dylib.0 -Wl,-framework,CoreFoundation -Wl,-framework,Carbon -Wl,-framework,Foundation -Wl,-framework,AppKit glib/libglib-2.0.0.dylib gobject/libgobject-2.0.0.dylib gmodule/libgmodule-2.0.0.dylib /nix/store/rfdg8nclvfhv4pcyqx0iaqk27173p0gm-zlib-1.2.11/lib/libz.dylib -Wl,--start-group -lresolv -Wl,--end-group -Wl,-headerpad_max_install_names -Wl,-rpath,@loader_path/../glib -Wl,-rpath,@loader_path/../gobject -Wl,-rpath,@loader_path/../gmodule -Wl,-rpath,/nix/store/y09iv91qmr970jz8yqg08vsjjf190phx-pcre-8.42/lib -Wl,-rpath,/nix/store/hq8zhb906blq4kw3q1p283yfzmyynhiy-libffi-3.2.1/lib -Wl,-rpath,/nix/store/rfdg8nclvfhv4pcyqx0iaqk27173p0gm-zlib-1.2.11/lib -Wl,-rpath-link,/private/tmp/nix-build-glib-2.58.2.drv-0/glib-2.58.2/build/glib:/nix/store/y09iv91qmr970jz8yqg08vsjjf190phx-pcre-8.42/lib:/private/tmp/nix-build-glib-2.58.2.drv-0/glib-2.58.2/build/gobject:/nix/store/hq8zhb906blq4kw3q1p283yfzmyynhiy-libffi-3.2.1/lib:/private/tmp/nix-build-glib-2.58.2.drv-0/glib-2.58.2/build/gmodule:/nix/store/rfdg8nclvfhv4pcyqx0iaqk27173p0gm-zlib-1.2.11/lib
ld: unknown option: -O1
clang-5.0: error: linker command failed with exit code 1 (use -v to see invocation)
```

The linker flag apparently comes from

https://github.com/mesonbuild/meson/blob/bbcf80d7348135e0590622a7b85c30429f005c07/mesonbuild/compilers/compilers.py#L177-L183

which is selected in

https://github.com/mesonbuild/meson/blob/bbcf80d7348135e0590622a7b85c30429f005c07/mesonbuild/compilers/compilers.py#L1477-L1479

The Objective-C compiler returns the following output for `--version`

```
clang version 5.0.2 (tags/RELEASE_502/final)
Target: x86_64-apple-darwin17.7.0
Thread model: posix
InstalledDir: /nix/store/frq6nffl9iszwlznlha23jyv2p1rrgc9-clang-5.0.2/bin
```

and when I pass it `-Wl,-v`, I get

```
@(#)PROGRAM:ld  PROJECT:ld64-274.2
configured to support archs: armv6 armv7 armv7s arm64 i386 x86_64 x86_64h armv6m armv7k armv7m armv7em (tvOS)
Library search paths:
        […]
Framework search paths:
        […]
Undefined symbols for architecture x86_64:
  "_main", referenced from:
     implicit entry/start for main executable
ld: symbol(s) not found for architecture x86_64
clang-5.0: error: linker command failed with exit code 1 (use -v to see invocation)
```

This means the compiler object will be instantiated with `CompilerType.CLANG_STANDARD` instead of `CompilerType.CLANG_OSX`

https://github.com/mesonbuild/meson/blob/0.49.0/mesonbuild/environment.py#L801

even though we use the Apple linker (actually ld64 from https://github.com/tpoechtrager/cctools-port).

C compiler works fine since it [detects](https://github.com/mesonbuild/meson/blob/0.49.0/mesonbuild/environment.py#L671) Darwin with `mesonlib.for_darwin(want_cross, self)`.

As a stopgap, we could add the same check to `detect_objc_compiler` but in the long term, Meson should probably check the linker name for determining whether `gnulike_buildtype_linker_args` should be used.

Encountered on Meson 0.49.0.

### LINKED ISSUE #4837: Separate shared linker and compiler representations
Our current abstraction assumes that one compiler will always use the same linker, GCC + GNU LD, Clang + GNU LD, cl.exe + link.exe etc. In some cases this is a pretty safe bet, but there are cases it doesn't work out. You might want ld.gold instead of ld.bfd, or ld.lld for clang. You might want to use the apple linker with GCC or upstream clang. You might need to use the solaris linker with gcc instead of gnu ld, etc.

I know I have a branch and I think @jon-turney has a branch to try to tease these apart. This bug was made to collect all of the issues that are blocking on this, and make sure we don't have 6 different people working on the same problem.

### LINKED ISSUE #5595: ld: fatal: unrecognized option '--as-needed' on Solaris
I'm using Meson 0.51 installed via pip. I followed the [Tutorial](https://mesonbuild.com/Tutorial.html).

```
$ cat meson.build 
project('tutorial', 'c')
executable('demo', 'main.c')
```

```
$ cat main.c 
#include<stdio.h>

int main(int argc, char **argv) {
  printf("Hello there.\n");
  return 0;
}
```

```
$ CC="gcc-6" meson builddir
The Meson build system
Version: 0.51.0
Source dir: /export/home/vedranm/workspace/demo
Build dir: /export/home/vedranm/workspace/demo/builddir
Build type: native build
Project name: tutorial
Project version: undefined
C compiler for the build machine: gcc-6 (gcc 6.5.0 "gcc-6 (OpenIndiana 6.5.0-OI-1) 6.5.0")
C compiler for the host machine: gcc-6 (gcc 6.5.0 "gcc-6 (OpenIndiana 6.5.0-OI-1) 6.5.0")
WARNING: Unknown CPU family 'i86pc', please report this at https://github.com/mesonbuild/meson/issues/new with theoutput of `uname -a` and `cat /proc/cpuinfo`
Build machine cpu family: i86pc
Build machine cpu: i86pc
Build targets in project: 1
Found ninja-1.8.2 at /usr/bin/ninja
```

```
$ cd builddir/
$ ninja 
[2/2] Linking target demo.
FAILED: demo 
gcc-6  -o demo 'demo@exe/main.c.o' -Wl,--no-undefined -Wl,--as-needed
ld: fatal: unrecognized option '--as-needed'
ld: fatal: use the -z help option for usage information
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

### PR DESCRIPTION
**Pull request by dcbaker**
This series is a start toward breaking the representations of the Compiler and Linker objects. It does this by creating a new `DynamicLinker` class and internally the compiler dispatches to it.

I have tested with all of the compilers I can, I need help from people with weird or proprietary compilers.

I have no way to test with a number of the proprietary compilers like PGI, ccrx, or cuda. I'll start pinging those people as we get closer to being able to merge this.

Fixes #4837 
Fixes #4784
Fixes #5595

### CODE REVIEW

**Comment by ignatenkobrain** [comment]
Flake8 detected 14 issues on 0ad3dc7fbc7cf940a3e0a18eeb783b6250d14a14
Visit https://sider.review/gh/repos/19784232/pulls/5681 to review the issues.

<sub>posted by Sider</sub>

**Comment by dcbaker** [comment]
Apple clang + the apple linker now works, real gcc (from homebrew) + the apple linker mostly works.

File: mesonbuild/compilers/compilers.py [hunk 1]
```diff
@@ -13,7 +13,9 @@
 # limitations under the License.
 
 import contextlib, enum, os.path, re, tempfile, shlex
-from typing import Optional, Tuple
+import typing
+from typing import List, Optional, Tuple
+from pathlib import Path
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/compilers.py]: __[Flake8]__ [F401] 'pathlib.Path' imported but unused ([link](https://sider.review/gh/repos/19784232/pulls/5681#issue-4358595))

<sub>posted by Sider</sub>

File: mesonbuild/compilers/mixins/islinker.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+    def get_undefined_link_args(self) -> typing.List[str]:
+        return []
+
+    def get_coverage_link_args(self) -> typing.List[str]:
+        m = "Linker {} doesn't implement coverage data generation.".format(self.id)
+        raise mesonlib.EnvironmentException(m)
+
+    def no_undefined_link_args(self) -> typing.List[str]:
+        return []
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/mixins/islinker.py]: __[Flake8]__ [W292] no newline at end of file ([link](https://sider.review/gh/repos/19784232/pulls/5681#issue-4358596))

<sub>posted by Sider</sub>

**Comment by dcbaker** [comment]
I now have ICC, GCC, Clang, and Apple Clang all working on macos along with the Linux compilers.

**Comment by ignatenkobrain** [comment]
Flake8 detected 7 issues on 590ee56cf5cebc85d5c7be469da664a162a24334
Visit https://sider.review/gh/repos/19784232/pulls/5681 to review the issues.

<sub>posted by Sider</sub>

**Comment by dcbaker** [comment]
I've taken a stab at getting PGI, the arm compilers and the ccrx compiler sorted. No idea if they actually work or not. I think I broke apple again.

File: mesonbuild/linkers.py [hunk 3]
```diff
@@ -12,249 +12,536 @@
...
+        'minsize': [],
+        'custom': [],
+    }  # type: typing.Dict[str, typing.List[str]]
+
+    def get_pie_args(self) -> typing.List[str]:
+        return ['-pie']
+
+    def get_asneeded_args(self) -> typing.List[str]:
+        return ['-Wl,--as-needed']
+
```
Comment (ignatenkobrain) [inline mesonbuild/linkers.py]: __[Flake8]__ [W293] blank line contains whitespace ([link](https://sider.review/gh/repos/19784232/pulls/5681#issue-4377405))

<sub>posted by Sider</sub>

File: mesonbuild/linkers.py [hunk 4]
```diff
@@ -12,249 +12,536 @@
...
+        return False
+
+    def get_std_shared_lib_args(self) -> typing.List[str]:
         return []
+
+
+class ArmClangDynamicLinker(ArmDynamicLinker):
+
+    """Linker for the linker used with ARM's clang fork.
+
```
Comment (ignatenkobrain) [inline mesonbuild/linkers.py]: __[Flake8]__ [W293] blank line contains whitespace ([link](https://sider.review/gh/repos/19784232/pulls/5681#issue-4377406))

<sub>posted by Sider</sub>

File: mesonbuild/compilers/mixins/islinker.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+    def bitcode_args(self) -> typing.List[str]:
+        raise mesonlib.MesonException("This linker doesn't support bitcode bundles")
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/mixins/islinker.py]: __[Flake8]__ [W292] no newline at end of file ([link](https://sider.review/gh/repos/19784232/pulls/5681#issue-4377408))

<sub>posted by Sider</sub>

File: mesonbuild/environment.py [hunk 6]
```diff
@@ -1085,10 +1127,10 @@ def detect_d_compiler(self, for_machine: MachineChoice):
...
                 raise EnvironmentException('Could not find any supported D compiler.')
 
-        try:
-            p, out = Popen_safe(exelist + ['--version'])[0:2]
-        except OSError:
-            raise EnvironmentException('Could not execute D compiler "%s"' % ' '.join(exelist))
+        #try:
+        p, out = Popen_safe(exelist + ['--version'])[0:2]
+        #except OSError:
+            #raise EnvironmentException('Could not execute D compiler "%s"' % ' '.join(exelist))
```
Comment (ignatenkobrain) [inline mesonbuild/environment.py]: __[Flake8]__ [E116] unexpected indentation (comment) ([link](https://sider.review/gh/repos/19784232/pulls/5681#issue-4377407))

<sub>posted by Sider</sub>

File: mesonbuild/linkers.py [hunk 7]
```diff
@@ -12,249 +12,538 @@
...
```
Comment (ignatenkobrain) [inline mesonbuild/linkers.py]: __[Flake8]__ [W293] blank line contains whitespace ([link](https://sider.review/gh/repos/19784232/pulls/5681#issue-4377416))

<sub>posted by Sider</sub>

File: mesonbuild/linkers.py [hunk 8]
```diff
@@ -12,249 +12,538 @@
...
+        'minsize': [],
+        'custom': [],
+    }  # type: typing.Dict[str, typing.List[str]]
+
+    def get_pie_args(self) -> typing.List[str]:
+        return ['-pie']
+
+    def get_asneeded_args(self) -> typing.List[str]:
+        return ['-Wl,--as-needed']
+
```
Comment (ignatenkobrain) [inline mesonbuild/linkers.py]: __[Flake8]__ [W293] blank line contains whitespace ([link](https://sider.review/gh/repos/19784232/pulls/5681#issue-4377415))

<sub>posted by Sider</sub>

File: mesonbuild/linkers.py [hunk 9]
```diff
@@ -12,249 +12,615 @@
...
+    def get_std_shared_lib_args(self) -> typing.List[str]:
+        return ['-shared']
+
+    def get_search_args(self, dirname: str) -> typing.List[str]:
+        return ['-L', dirname]
+
+
+
+class GnuLikeDynamicLinkerMixin:
```
Comment (ignatenkobrain) [inline mesonbuild/linkers.py]: __[Flake8]__ [E303] too many blank lines (3) ([link](https://sider.review/gh/repos/19784232/pulls/5681#issue-4377436))

<sub>posted by Sider</sub>

**Comment by ignatenkobrain** [comment]
Flake8 detected 6 issues on 6c8292a9d33324aebf0aaa21c01c3a859c5c0165
Visit https://sider.review/gh/repos/19784232/pulls/5681 to review the issues.

<sub>posted by Sider</sub>

**Comment by dcbaker** [comment]
I've introduced a few new bugs on Linux, but I have cl.exe/link.exe and clang-cl/ldd-link.exe working, icl/xilink.exe is kinda working, but I don't konw if that's from this series or previous breakage. No idea on Apple. I've made an attempt at cuda, but it's blind since cuda requires the Nviida drivers to even install. I've realized the ccrx linker implementation I've done is completely bogus. That's all for today.

**Comment by ignatenkobrain** [comment]
Flake8 detected 9 issues on 2a69fd5490e06bba3cc2806daac97852a2ed09fc
Visit https://sider.review/gh/repos/19784232/pulls/5681 to review the issues.

<sub>posted by Sider</sub>

File: mesonbuild/compilers/mixins/islinker.py [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+"""
+
+import os
+import shlex
+import typing
+
+from ... import mesonlib
+
+if typing.TYPE_CHECKING:
+    from ...coredata import OptionDictType
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/mixins/islinker.py]: __[Flake8]__ [W291] trailing whitespace ([link](https://sider.review/gh/repos/19784232/pulls/5681#issue-4389284))

<sub>posted by Sider</sub>

**Comment by ignatenkobrain** [comment]
Flake8 detected 10 issues on 91d517f861f9fe00eed9d5eeaba652805098cb24
Visit https://sider.review/gh/repos/19784232/pulls/5681 to review the issues.

<sub>posted by Sider</sub>

**Comment by dcbaker** [comment]
Getting very, very close to getting all of the compilers I have access to working at least as well as they did before (There are still some problems with using gcc and non-apple clang on mac, but it's still better than the current status of "nothing works at all")

**Comment by dcbaker** [comment]
@obilaniu I don't know who to ask about cuda, but you've done a bunch of work on it, so maybe you'd be willing to help?

@ftechz, I know that the ccrx support is still really busted but I at least wanted to make you aware of this.

@scivision, @mohdamerkhalidi @sompen @mbedarka @makise-homura 

If I could ask you guys to take a look at the compiler's I've tried to get working (PGI, CCRX, Cuda, LCC, ArmCc, and ArmClang), but I either can't get access to or have no way to run that would be fantastic. If things aren't working a list of failing tests and backtraces would be useful. If you're feeling particularly generous and want to send me patches I'd very much appreciate that.

**Comment by obilaniu** [comment]
@dcbaker Yeah, there are some blowups. I'm running with CUDA Toolkit 10.1 and a Kepler GPU, and I get a few backtraces. The minimum patch to allow the testcases to run is:

```diff
diff --git a/mesonbuild/compilers/cuda.py b/mesonbuild/compilers/cuda.py
index a0019bbe..54d2a453 100644
--- a/mesonbuild/compilers/cuda.py
+++ b/mesonbuild/compilers/cuda.py
@@ -166,7 +166,8 @@ class CudaCompiler(Compiler):
         Converts GNU-style arguments -Wl,-arg,-arg
         to NVCC-style arguments -Xlinker=-arg,-arg
         """
-        return [arg.replace('-Wl', '-Xlinker=', 1) if arg.startswith('-Wl') else arg for arg in args]
+        args = [arg.replace('-Wl', '-Xlinker=', 1) if arg.startswith('-Wl') else arg for arg in args]
+        return [arg.replace(' ', '\\') for arg in args]
 
     def name_string(self):
         return ' '.join(self.exelist)
@@ -189,6 +190,9 @@ class CudaCompiler(Compiler):
     def get_debug_args(self, is_debug):
         return cuda_debug_args[is_debug]
 
+    def get_output_args(self, target):
+        return ['-o', target]
+
     def get_werror_args(self):
         return ['-Werror=cross-execution-space-call,deprecated-declarations,reorder']
 
@@ -212,8 +216,8 @@ class CudaCompiler(Compiler):
     def build_rpath_args(self, env: 'Environment', build_dir: str, from_dir: str,
                          rpath_paths: str, build_rpath: str,
                          install_rpath: str) -> typing.List[str]:
-        return self._cook_link_args(super().build_rpath_args(
-            build_dir, from_dir, rpath_paths, build_rpath, install_rpath))
+        return self._cook_link_args(self.linker.build_rpath_args(
+            env, build_dir, from_dir, rpath_paths, build_rpath, install_rpath))
 
     def linker_to_compiler_args(self, args):
         return args
@@ -223,3 +227,6 @@ class CudaCompiler(Compiler):
 
     def compute_parameters_with_absolute_paths(self, parameter_list, build_dir):
         return []
+
+    def get_std_exe_link_args(self):
+        return []
diff --git a/mesonbuild/environment.py b/mesonbuild/environment.py
index d16ce480..a1add802 100644
--- a/mesonbuild/environment.py
+++ b/mesonbuild/environment.py
@@ -912,7 +912,7 @@ class Environment:
             # Luckily, the "V" also makes it very simple to extract
             # the full version:
             version = out.strip().split('V')[-1]
-            linker = self._guess_nix_linker(compiler, for_machine, prefix=['-Xlinker='])
+            linker = self._guess_nix_linker(compiler, for_machine, prefix='-Xlinker=')
             return CudaCompiler(ccache + compiler, version, for_machine, exe_wrap, linker=linker)
         raise EnvironmentException('Could not find suitable CUDA compiler: "' + ' '.join(compilers) + '"')
 
```

I also get for every single build, CUDA or not, a DEPRECATION warning that seems spurious:

```
The Meson build system
Version: 0.51.0
Source dir: /Meson/test cases/common/99 threads
Build dir: /Meson/test cases/common/99 threads/build
Build type: native build
DEPRECATION: Duplicated values in array option is deprecated. This will become a hard error in the future.
Project name: threads
Project version: undefined
C compiler for the build machine: cc (gcc 7.4.0 "cc (SUSE Linux) 7.4.0")
C++ compiler for the build machine: c++ (gcc 7.4.0 "c++ (SUSE Linux) 7.4.0")
C compiler for the host machine: cc (gcc 7.4.0 "cc (SUSE Linux) 7.4.0")
C++ compiler for the host machine: c++ (gcc 7.4.0 "c++ (SUSE Linux) 7.4.0")
Build machine cpu family: x86_64
Build machine cpu: x86_64
Run-time dependency threads found: YES 
Build targets in project: 2
Found ninja-1.8.2 at /usr/bin/ninja
[4/4] Linking target cppthreadprog.
```

This message appeared somewhere between 0.50.1 and 0.51.0 though.

**Comment by scivision** [comment]
Re: PGI, this was perhaps my first PR to Meson--I didn't make sure that all the tests passed, just that all my various projects worked. So in `master`, probably since I "added" PGI, there are several project tests broken.

One idea is to have a branch off of master just to fix PGI. There will be unique issues for PGI on Windows since PGI doesn't have it's own C++ compiler on Windows.
Maybe a strategy for now is just to list the tests that are broken with PGI, and see that this PR #5681 doesn't break anything additional?

**Comment by scivision** [comment]
Re: PGI, I made a Gist of Linux tests that fail for PGI on Meson `master`. https://gist.github.com/scivision/6b6c7833751d9f69b340d594459dbd26

Then I made a PR fixing common/13 pch test: #5731

**Comment by dcbaker** [comment]
@obilaniu. Thanks for the cuda changes, I'll integrate those. RE: the duplicates warning. It's because we've added a command line argument for pkg_config_path, and the environment is used if the command line argument is not passed, so that's a bad warning in that case. Could you open a bug about it and mention me, I'll fix it.

File: mesonbuild/compilers/cuda.py [hunk 11]
```diff
@@ -213,7 +219,7 @@ def build_rpath_args(self, env: 'Environment', build_dir: str, from_dir: str,
                          rpath_paths: str, build_rpath: str,
                          install_rpath: str) -> typing.List[str]:
         return self._cook_link_args(super().build_rpath_args(
-            build_dir, from_dir, rpath_paths, build_rpath, install_rpath))
```
Comment (obilaniu) [inline mesonbuild/compilers/cuda.py]: @dcbaker There was one other important change to this line - I switched out `super().` for `self.linker.`

Comment (dcbaker) [reply mesonbuild/compilers/cuda.py]: Did we actually need to replace the super() call, or does adding the missing `env` variable fix it for you? I'd prefer not to override the super() call to do exactly what the super call is supposed to do.

Comment (obilaniu) [reply mesonbuild/compilers/cuda.py]: @dcbaker Just the `env` var seems to work, **but** you have accidentally introduced a small typo:

`get_std_exe_link_arg` -> `get_std_exe_link_args`

Comment (dcbaker) [reply mesonbuild/compilers/cuda.py]: oops, good catch. Updated to fix that.

**Comment by jpakkane** [comment]
Trying to run this with the ARM compiler under Linux produces the error "armlink.exe not found". Then again, master does the same. This is probably because armcc has only been used on Windows thus far.

**Comment by dcbaker** [comment]
@jpakkane, I vaguely remember that when the arm compilers were enabled that it was only working for windows. I guess we could try removing the `.exe` and see what happens. Thanks for looking though.

**Comment by dcbaker** [comment]
I'll get the D thing sorted today. I've installed a debian chroot so I can get a working GDC (it's still broken on ARCH...)

File: mesonbuild/compilers/mixins/gnu.py [hunk 12]
```diff
@@ -380,10 +278,19 @@ def sanitizer_compile_args(self, value: str) -> typing.List[str]:
...
-        return ['-fsanitize=' + value]
+    def get_output_args(self, target: str) -> typing.List[str]:
+        return ['-o', target]
+
+    def get_dependency_gen_args(self, outtarget, outfile):
+        return ['-MD', '-MQ', outtarget, '-MF', outfile]
+
+    def get_compile_only_args(self) -> typing.List[str]:
+        return ['-c']
+
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/mixins/gnu.py]: __[Flake8]__ [W293] blank line contains whitespace ([link](https://sider.review/gh/repos/19784232/pulls/5681#issue-4401266))

<sub>posted by Sider</sub>

**Comment by dcbaker** [comment]
@ftechz @mohdamerkhalidi @sompen @mbedarka @makise-homura; All of the compiler/linker combinations I have access to are now working. Can you have a look at the compilers you've enabled and see if things are working? I would really appreciate patches if possible.

@scivision, what's the status of PGI now.

**Comment by scivision** [comment]
@dcbaker there is something wrong for PGI, even common/1 trivial immediately errors

> ERROR: Linker pgi does not support allow undefined

**Comment by scivision** [comment]
@dcbaker fixes for those PGI issues in https://github.com/scivision/meson/tree/dcbaker-dynamic-linker-split

**Comment by scivision** [comment]
However, the Fortran tests all fail as follows. A straightforward fix perhaps?

> Traceback (most recent call last):
  File "\mesonbuild\mesonmain.py", line 129, in run
    return options.run_func(options)
  File "\mesonbuild\msetup.py", line 241, in run
    app.generate()
  File "\mesonbuild\msetup.py", line 159, in generate
    self._generate(env)
  File "\mesonbuild\msetup.py", line 175, in _generate
    intr = interpreter.Interpreter(b)
  File "\mesonbuild\interpreter.py", line 2088, in __init__
    self.parse_project()
  File "\mesonbuild\interpreterbase.py", line 397, in parse_project
    self.evaluate_codeblock(self.ast, end=1)
  File "\mesonbuild\interpreterbase.py", line 436, in evaluate_codeblock
    raise e
  File "\mesonbuild\interpreterbase.py", line 430, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "\mesonbuild\interpreterbase.py", line 441, in evaluate_statement
    return self.function_call(cur)
  File "\mesonbuild\interpreterbase.py", line 778, in function_call
    return func(node, posargs, kwargs)
  File "\mesonbuild\interpreterbase.py", line 143, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "\mesonbuild\interpreterbase.py", line 174, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "\mesonbuild\interpreter.py", line 2737, in func_project
    self.add_languages(proj_langs, True)
  File "\mesonbuild\interpreter.py", line 2796, in add_languages
    success = self.add_languages_for(args, required, MachineChoice.BUILD)
  File "\mesonbuild\interpreter.py", line 2810, in add_languages_for
    comp = self.environment.detect_compiler_for(lang, for_machine)
  File "\mesonbuild\environment.py", line 1236, in detect_compiler_for
    self.coredata.process_new_compiler(lang, comp, self)
  File "\mesonbuild\coredata.py", line 720, in process_new_compiler
    for k, o in comp.get_and_default_options(env.properties[comp.for_machine]).items():
  File "\mesonbuild\compilers\compilers.py", line 843, in get_and_default_options
    compile_args, link_args = self.get_args_from_envvars()
  File "\mesonbuild\compilers\compilers.py", line 800, in get_args_from_envvars
    env_link_flags = self.get_linker_args_from_envvars()
  File "\mesonbuild\compilers\compilers.py", line 771, in get_linker_args_from_envvars
    return self.linker.get_args_from_envvars()
AttributeError: 'NoneType' object has no attribute 'get_args_from_envvars'

**Comment by scivision** [comment]
...Flang Fortran tests work

**Comment by scivision** [comment]
on INtel for Linux, the Fortran tests all fail. On intel for Windows they're OK

**Comment by dcbaker** [comment]
@scivision, the ifort thing is a bug that's always existed manifesting in a strange way. It turns out we've never generated .d files with ifort on mac and Linux, I added the GNU interface for that to the GnuLikeMixin which ifort doesn't support. I've fixed that.

I've also pulled your branch, squashed all of the linker commits into the "Add PGI linker" commit, and re-authored that to you (since you wrote all of the code). I've cleaned up the cleanup commits a bit and merged them into the early cleanup part of the series. I've also fixed (I think) the pgi fortran problem, which was not adding a dynamic linker instance for it. I'm still working out a couple of rebase bugs and then I'll push again.

File: mesonbuild/interpreter.py [hunk 13]
```diff
@@ -2824,7 +2824,10 @@ def add_languages_for(self, args, required, for_machine: MachineChoice):
                         raise
             if machine_name != 'build' or self.environment.is_cross_build():
                 mlog.log(comp.get_display_language(), 'compiler for the', machine_name, 'machine:',
-                         mlog.bold(' '.join(comp.get_exelist())), comp.get_version_string())
+                        mlog.bold(' '.join(comp.get_exelist())), comp.get_version_string())
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[Flake8]__ [E128] continuation line under-indented for visual indent ([link](https://sider.review/gh/repos/19784232/pulls/5681#issue-4423135))

<sub>posted by Sider</sub>

File: mesonbuild/interpreter.py [hunk 14]
```diff
@@ -2824,7 +2824,10 @@ def add_languages_for(self, args, required, for_machine: MachineChoice):
...
+                if comp.linker is not None:
+                    mlog.log(comp.get_display_language(), 'linker for the', machine_name, 'machine:',
+                            mlog.bold(comp.linker.id), comp.linker.version)
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[Flake8]__ [E128] continuation line under-indented for visual indent ([link](https://sider.review/gh/repos/19784232/pulls/5681#issue-4423136))

<sub>posted by Sider</sub>

**Comment by dcbaker** [comment]
@ftechz @mohdamerkhalidi @sompen @mbedarka @makise-homura; ping

@jpakkane, I saw you enabled armcc on linux, could you try again and see if this works for you?

**Comment by dcbaker** [comment]
I've updated emscripten to use the `CompilerIsLinker` infrastructure, and fixed a couple of missed methods in the `Ccrx` family that were shadowing the same methods in the their linker. I still really need to hear back from @ftechz @mohdamerkhalidi @sompen @mbedarka @makise-homura
@jpakkane about whether the compilers/linkers they added/have access to work.

**Comment by obilaniu** [comment]
@dcbaker It has been two weeks since you first pinged us to verify the more exotic compilers work, and only PGI and NVCC were stepped up. The others are: ARMCC/ARMClang, which ship with Keil, an already-complete development kit; and lcc targetting Elbrus CPUs, for which no CI system is available because they are nearly exclusive to the Moscow Center for SPARC Technologies (MCST).

It may be reasonable to obsolete them in 0.52.0 and drop them in 0.53.0 unless ARM, MCST or volunteers step up, and institute some type of loose "maintainership" of the various compilers.

For my part, I'm willing to maintain for some time the CUDA module and keep NVCC running, although currently the GPU project I have retrieves `nvcc` manually and uses it in `custom_target()`s to build CUDA kernels. The reason why mostly has to do with the painful variety of different ways the CUDA Toolkit is installed, officially and unofficially, on many Linux distributions and HPC clusters that exist.

File: mesonbuild/compilers/compilers.py [hunk 15]
```diff
@@ -335,18 +341,26 @@ def get_base_link_args(options, linker, is_shared_module):
...
     # -bitcode_bundle is incompatible with -undefined and -bundle
     if bitcode and not is_shared_module:
-        args.append('-Wl,-bitcode_bundle')
+        args.extend(linker.bitcode_args())
     elif as_needed:
         # -Wl,-dead_strip_dylibs is incompatible with bitcode
         args.extend(linker.get_asneeded_args())
+
+    # Apple's ld (the only one that supports bitcode) does not like any
+    # -undefined arguments at all, so don't pass these when using bitcode
```
Comment (ftechz) [inline mesonbuild/compilers/compilers.py]: `ccrx` does not support these arguments as well. Any way to bypass it as well?

Comment (ftechz) [reply mesonbuild/compilers/compilers.py]: I'm getting this error `Linker rlink does not support allow undefined`

**Comment by jpakkane** [comment]
ARM on Linux fails with:

    ERROR: armlink does not support allow undefined

As a reminder I do have license keys for ARM Linux compilers, which I can share with dev team members if they want to run the tests themselves.

**Comment by sompen** [comment]
@dcbaker Apologies for the delay.
I have checked ArmCc and ArmClang compilers with my local test setup and the tests were initially failing due to missing methods - get_allow_undefined_args and get_soname_args, for Arm family. After overriding the two methods for Arm family, the setup works fine.

Patch snippet for Arm family compiler/linker failure fixes:
```
diff --git a/mesonbuild/linkers.py b/mesonbuild/linkers.py
index 9f0d7309..a0e58d2c 100644
--- a/mesonbuild/linkers.py
+++ b/mesonbuild/linkers.py
@@ -677,6 +677,14 @@ class ArmDynamicLinker(PosixDynamicLinkerMixin, DynamicLinker):
     def get_std_shared_lib_args(self) -> typing.List[str]:
         return []

+    def get_allow_undefined_args(self) -> typing.List[str]:
+        return []
+
+    def get_soname_args(self, env: 'Environment', prefix: str, shlib_name: str,
+                        suffix: str, soversion: str, darwin_versions: typing.Tuple[str, str],
+                        is_shared_module: bool) -> typing.List[str]:
+        return []
+
```

File: mesonbuild/compilers/compilers.py (hunk 15 above)
Comment (dcbaker) [reply mesonbuild/compilers/compilers.py]: Apple does support `-undefined,<foo>`, just no in conjunction with `-bitcode-bundle`.

I think we just need to implement a stub function.

**Comment by dcbaker** [comment]
@sopen, thanks! I've integrated your changes for the arm linkers.

@ftechz, I've added the same changes to the rlink that I added to the arm linker, I think might fix all of it, but at least should get us further along.

**Comment by dcbaker** [comment]
I've also added an implementation of the illumos dynamic linker, I'd originally planned to do this separately but this is taking so long I just added it here.

**Comment by jpakkane** [comment]
For Armclang executables work but shared libraries error out with `ArmClangdynamicLinker` missing `get_soname_args`. In current master it works. Or at least compiles, I don't know if the Arm compiler can really do share libraries, because according to `file` the files it outputs are static linked rather than shared objects.

The linker also complains about missing entry points when creating the shared libraries. Which probably means that shared linking is not supported. But it would be great if it did not fail with a traceback at least.

**Comment by dcbaker** [comment]
@jpakkane ,something like that?

Also, I'm looking into the windows/apple regression.

**Comment by jpakkane** [comment]
Seems reasonable.

**Comment by dcbaker** [comment]
It looks like msys is just failing for everyone?

File: mesonbuild/compilers/compilers.py (hunk 15 above)
Comment (dcbaker) [reply mesonbuild/compilers/compilers.py]: Can you try again? I think I have all of the necessary functions implemented now.

**Comment by dcbaker** [comment]
@makise-homura, any chance you can give this a whirl?

**Comment by dcbaker** [comment]
@ftechz, does the updated version work?

**Comment by ftechz** [comment]
> @ftechz, does the updated version work?

@dcbaker Sorry for the delayed reply. Yes it is working now 👍

**Comment by dcbaker** [comment]
@jpakkane, are we ready to merge?

**Comment by jpakkane** [comment]
CI had frozen so I restarted it.

**Comment by jpakkane** [comment]
And it's done. Thanks for getting this done.

**Comment by dcbaker** [comment]
I am glad to be done with it!

**Pull request merged**